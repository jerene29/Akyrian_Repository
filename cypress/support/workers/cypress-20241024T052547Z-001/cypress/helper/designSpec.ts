import { GetStudyRevisionListDocument } from '../../src/graphQL/generated/graphql';
import { mockUserDataAdmin } from '../../src/constant/testFixtures';

type SpecNumber = `S-${number}-${number}`;

type SpecInfo = {
  // e.g S-09-010
  spec: SpecNumber;
  // title e.g headerSection, combined description becomes S-09-010 headerSection, but optional
  title?: string;
};

type Options = {
  loginAs?: string;
  skipTest?: boolean;
};

function resolveDescription(description: SpecNumber | SpecInfo) {
  return typeof description === 'string' ? description : `${description.spec} ${description.title}`;
}

const aliasGetStudyRevisionListDocument =
  'name' in GetStudyRevisionListDocument.definitions[0]
    ? GetStudyRevisionListDocument.definitions[0].name?.value
    : 'GetStudyRevisionListDocument';
/*
example usage
designSpec('S-19-2', () => {

  describe('...', ...)

  it('checks something...', () => {...})
});
*/
export function designSpec(
  description: SpecNumber | SpecInfo,
  callback: () => void,
  options?: Options,
) {
  before(() => {
    cy.intercept('POST', '/graphql', (req) => {
      if (req.body.operationName === aliasGetStudyRevisionListDocument) {
        req.alias = req.body.operationName;
      }
    });
    cy.beforeSetup(mockUserDataAdmin);
    cy.clearLocalStorageSnapshot();
    cy.saveLocalStorage();
    cy.wait(`@${aliasGetStudyRevisionListDocument}`);
  });
  if (options?.skipTest) {
    describe.skip(resolveDescription(description), callback);
  } else if (typeof description !== 'string' && description.title) {
    describe(description.spec, () => {
      describe(String(description.title), callback);
    });
  } else {
    describe(resolveDescription(description), callback);
  }
}
