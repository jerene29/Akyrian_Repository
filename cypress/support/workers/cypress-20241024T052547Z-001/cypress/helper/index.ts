export * from './designSpec';
export * from './object';