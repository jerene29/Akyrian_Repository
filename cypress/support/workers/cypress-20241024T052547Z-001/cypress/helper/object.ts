/**
 * to shorten the tedious '[data-cy=selector]' selector.
 * just put the name of data-cy attributed element.
 * -- d stands for document
 *
```
d`addButton`
```
 */
export function d(args: TemplateStringsArray) {
  return `[data-cy=${args.join('')}]`;
}
