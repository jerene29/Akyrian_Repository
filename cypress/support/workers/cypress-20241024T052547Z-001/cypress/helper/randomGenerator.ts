import moment = require("moment");

export const randomAlphaNumeric = (length: number) => {
  const chars = '0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomAlphabet = (length: number) => {
  const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
  let result = '';
  for (let i = length; i > 0; --i) result += chars[Math.floor(Math.random() * chars.length)];
  return result;
}

export const randomDate = () => {
  const start = new Date(2012, 0, 1)
  const end = new Date()
  const date = new Date(start.getTime() + Math.random() * (end.getTime() - start.getTime()))
  return moment(date);
}