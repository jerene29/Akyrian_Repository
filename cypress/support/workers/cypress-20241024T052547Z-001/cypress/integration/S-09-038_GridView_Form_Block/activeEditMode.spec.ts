import { d, designSpec } from '../../helper';

const options = {
  loginAs: 'admin@example.com',
};

designSpec(
  {
    spec: 'S-09-038',
    title: 'Active Edit Mode on Form',
  },
  () => {
    before(() => {
      cy.createStudy({
        name: 'GRIDVIEW_TEST',
        // remember the initial form was 5
        formCount: 5,
        visitCount: 5,
      });
    });

    context('validates form name', () => {
      before(() => {
        cy.hover('#formHeader-1-0');
      });

      const nameInput = d`activeEditableFormName-1-0`;

      it('checks for blank field', () => {
        cy.shouldBeVisible(d`editableFormName-1-0`).click();
        cy.shouldBeVisible(nameInput).clear();
        cy.shouldBeVisible(d`cellErrorPopover`).getTextSnapshot();
      });

      it('checks alphanumeric validation', () => {
        cy.shouldBeVisible(nameInput).type('..?');
        cy.shouldBeVisible(d`cellErrorPopover`).getTextSnapshot();
      });

      it('must contains non space', () => {
        cy.shouldBeVisible(nameInput).clear().type('       ');
        cy.shouldBeVisible(d`cellErrorPopover`).getTextSnapshot();
      });

      it('checks the limit of the characters, max is 35', () => {
        cy.shouldBeVisible(nameInput)
          .clear()
          .type('awkawkakwawkawkwaokwaooawkeoawkokaowaowkaokwoawkaowaowkaokwoao');
        cy.shouldBeVisible(d`cellErrorPopover`).getTextSnapshot();
      });

      it('can cancels the edit', () => {
        cy.shouldBeVisible(nameInput).clear().type('lets cancel');
        cy.shouldBeVisible(d`editCancel`).click();
      });
    });

    context(
      'tab sequence - the #.(configurable parameters of a visit block) have a distinct tab order, the [TAB] button advances downwards, the [SHIFT + TAB] advances upwards, the order is as follows:',
      () => {
        before(() => {
          cy.hover('#formHeader-1-0');
        });

        // TODO: rework on this intermittently fail on cloud.
        xit('checks if pressing [tab] will advance beyond the current GV Form Block and open the next #.(form name) of the next GV Form Block (below)', () => {
          const EDITED_NAME = 'Edited Form 2';

          cy.get('body').tab();
          cy.shouldBeVisible(d`activeEditableFormName-2-0`)
            .focus()
            .clear()
            .type(EDITED_NAME);
          cy.get('body').tab();

          cy.get('#formHeader-4-0').realHover();
          cy.get('#formHeader-2-0').realHover();
          cy.get('#formHeader-2-0').contains(EDITED_NAME);
        });

        it('pressing [shift-tab] will advance beyond the current GV Form Block and open the prior #.(form name) of the next GV Form Block (above)', () => {
          cy.get('body').tab({ shift: true });
          cy.shouldBeVisible(d`editableFormName-1-0`);
        });
      },
    );
  },
  options,
);
