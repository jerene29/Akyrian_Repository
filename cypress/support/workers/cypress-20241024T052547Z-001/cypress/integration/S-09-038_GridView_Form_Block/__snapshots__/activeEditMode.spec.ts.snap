exports[`S-09-038 > Active Edit Mode on Form > validates form name > checks for blank field #0`] =
{
  "snapshot": "this field cannot be left blank"
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > checks for blank field #1`] =
{
  "snapshot": ""
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > checks alphanumeric validation #0`] =
{
  "snapshot": "can only contain letters, numbers and special characters: [space], -, _, #"
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > checks alphanumeric validation #1`] =
{
  "snapshot": ""
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > must contains non space #0`] =
{
  "snapshot": "you must enter at least one non-space character"
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > must contains non space #1`] =
{
  "snapshot": ""
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > checks the limit of the characters, max is 35 #0`] =
{
  "snapshot": "min 1 character, max 35 characters"
};

exports[`S-09-038 > Active Edit Mode on Form > validates form name > checks the limit of the characters, max is 35 #1`] =
{
  "snapshot": ""
};
